django/db/models/operators.py

```diff
@@ -0,0 +1,20 @@
+from django.db.models.expressions import CombinedExpression
+from django.db.models.fields import (
+    DateField, DateTimeField, DurationField, TimeField
+)
+
+__all__ = ()
+
+
+@CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateField)
+@CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateTimeField)
+@CombinedExpression.dispatch(CombinedExpression.SUB, TimeField, TimeField)
```

Comment:
Resurrecting this half a year later: IMHO most subclasses should work by default except a few known cases. For example I'd expect MyCustomDateField - DateField to just work but expect subclasses of DateTimeField to fail when combined with subclasses of DateField.

Stopping on first match is actually enough to handle this now but may be overly verbose:

```python
def handle(...):
    pass

def reject(...):
    raise NotImplementedError(...)

CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateTimeField)(handle)

CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateField)(reject)
CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateTimeField)(reject)

CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateField)(handle)
```

Possible solution would be to explicitly mark unhandled cases (keep in mind this would be mostly used internally by Django's built in set of fields):

```python
@CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateTimeField)
@CombinedExpression.dispatch(CombinedExpression.SUB, DateTimeField, DateField, unhandled=True)
@CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateTimeField, unhandled=True)
@CombinedExpression.dispatch(CombinedExpression.SUB, DateField, DateField)
def handle(...):
    pass
```

An easier fix would be to introduce a base class for the common part of DateField and DateTimeField (and other cases where it's convenient to share code but the types don't really make sense when combined) but it'd likely break backwards compatibility.